src/metabase/api/collection.clj

```diff
@@ -46,19 +46,56 @@
-                                     (remove (comp #{user-id} :personal_owner_id))
-                                     (map :id))
-                           collections)
-        prefixes     (into #{} (map (fn [id] (format "/%d/" id))) personal-ids)
-        personal?    (fn [{^String location :location id :id}]
-                       (or (personal-ids id)
-                           (prefixes (re-find #"^/\d+/" location))))]
-    (if (seq prefixes)
-      (remove personal? collections)
-      collections)))
+(defn- location-from-collection-id-clause
+  "Clause to restrict which collections are being selected based off collection-id. If collection-id is nil,
+   then restrict to the children and the grandchildren of the root collection. If collection-id is an an integer,
+   then restrict to that collection's parents and children."
+  [collection-id]
+  (if collection-id
+    [:and
+     [:like :location (str "%/" collection-id "/%")]
+     [:not [:like :location (str "%/" collection-id "/%/%/%")]]]
+    [:not [:like :location "/%/%/"]]))
+
+(defn- exclude-other-user-collections-clause
+  [user-id]
+  [:or [:= user-id :personal_owner_id] [:= nil :personal_owner_id]])
+
+(defn- remove-orphaned-collections
+  "Removes collections without it's root parent, used mainly for filtering out other user's
+   subcollections."
+  [collections]
+  (let [prefixes      (into #{}
+                            (map
+                             (fn [{:keys [id]}] (format "/%d/" id))
+                             collections))
+        parent-exists (fn [{:keys [location]}]
+                        (or (= "/" location)
+                            (prefixes (re-find #"^/\d+/" location))))]
+    (filter parent-exists collections)))
+
+(defn- select-collections
+  "Select collections based off certain parameters. If `shallow` is true, we select only the requested collection (or
+  the root, if `collection-id` is `nil`) and its immediate children, to avoid reading the entire collection tree when it
+  is not necessary."
+  [exclude-archived exclude-other-user-collections namespace shallow collection-id]
+  (cond->>
+   (t2/select Collection
+              {:where [:and
+                       (when exclude-archived
+                         [:= :archived false])
+                       (when shallow
+                         (location-from-collection-id-clause collection-id))
+                       (when exclude-other-user-collections
+                         (exclude-other-user-collections-clause api/*current-user-id*))
+                       (perms/audit-namespace-clause :namespace namespace)
+                       (collection/visible-collection-ids->honeysql-filter-clause
+                        :id
+                        (collection/permissions-set->visible-collection-ids @api/*current-user-permissions-set*))]
+               ;; Order NULL collection types first so that audit collections are last
```

Comment:
To consolidate other SQL statements, I added the order-by found in the "/" API request to both. This shouldn't functionally change the "/tree" request, and I don't think it would introduce much latency.